### ISSUE TITLE
Restriction Lint for static / const arrays filled with a non-zero value

### ISSUE THREAD

**Issue by llogiq** [rust-lang/rust-clippy#14805]
### What it does

Thanks to Cliff Biffle for bringing this up during RustWeek '25.

### Advantage

In embedded firmware, a nonzero static array takes up the size of the array in flash, whereas a zero-initialized array + fill(_) is only a few bytes.

### Drawbacks

There should probably be some cutoff size below which the lint should not trigger.

### Example

```rust
static FOO: &[u8; 1024] = [42; 1024];
```

should be written as

```rust
static mut FOO: &[u8; 1024] = [0; 1024];

unsafe { FOO.fill(42) };
```

**Comment by samueltardieu** [comment]
This would be easier to do in languages such as C++ where constructors for global variables run before the main program starts. While I perfectly understand the motivation (don't stuff `.rodata`, use `.bss` instead), finding a suggestion which doesn't make the compiler complain because of `static mut` might prove difficult to find (the `static_mut_refs` is now deny-by-default in edition 2024, while it was "only" warn-by-default in edition 2021 and earlier).

Another consideration would be that, quite often in embedded systems, RAM is more limited then flash, so using some `.rodata` bytes might be less problematic than using the equivalent in RAM.

Maybe this lint should only warn, reminding the user that building a non-zero constant array will consume space in `.rodata`, and that it would be preferable to find another solution.

**Comment by llogiq** [comment]
Yeah, there is currently some discussion going on around `static mut`. The opsem people want it to go away, and the embedded folks have no alternative.

**Comment by llogiq** [comment]
(Oops, clicked the wrong button)
